### PR TITLE
SYCL TeamPolicy scratch memory

### PR DESCRIPTION
**Pull request by masterleinad**
Based on top of https://github.com/kokkos/kokkos/pull/3654.
Currently, I don't see a way to enable thread scratch memory so I disabled the corresponding incremental test for now.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
The problem I am seeing with implementing thread scratch memory is that the interface we need to support looks somewhat like https://github.com/kokkos/kokkos/blob/78538c8574777342f64c7da657f419778eaa2a10/core/unit_test/incremental/Test12a_ThreadScratch.hpp#L67-L85
In particular, `team.thread_scratch()` is called within the outer TeamPolicy `parallel_for` loop which maps to `parallel_for_workgroup` and the information about the active team essentially comes from the corresponding `group` instance. Before SYCL 2020, i.e. with SYCL 1.2.1 (https://www.khronos.org/registry/SYCL/specs/sycl-1.2.1.pdf) the `group` class doesn't have a `get_local_id()` member function that would correspond to our `team_rank()`. Hence, I don't see a way for `thread_scratch()` to do anything sensible.

The appropriate SYCL concept looks like
```c++
myQueue.submit([&](handler & cgh) {
  cgh.parallel_for_work_group<class example_kernel>(
      range<3>(2, 2, 2), range<3>(2, 2, 2), [=](group<3> myGroup) {

    //[workgroup code]
    private_memory<int> myPrivate(myGroup);

    myGroup.parallel_for_work_item([&](h_item<3> myItem) {
      //[work-item code]
      myPrivate(myItem) = 0;
    });
    //[workgroup code]
  });
});
```
which is similar but not quite the same to what we do since here the memory to be accessed is only specified in the inner `parallel_for_work_item` loop.

**Comment by masterleinad** [comment]
Depends on #3654.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp [hunk 1]
```diff
@@ -441,7 +446,34 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
         m_league_size(arg_policy.league_size()),
         m_team_size(arg_policy.team_size()),
         m_vector_size(arg_policy.impl_vector_length()) {
-    // FIXME_SYCL check_parameters
+    m_shmem_size =
+        (m_policy.scratch_size(0, m_team_size) +
+         m_policy.scratch_size(1, m_team_size) +
+         FunctorTeamShmemSize<FunctorType>::value(m_functor, m_team_size));
```
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:452]: Wait are you trying to get both lvl 1 and lvl 0 scratch from the actual sycl shmem which is likely limited by GPU shmem? That is not the intend. I'd rather have you say lvl 1 max size is zero instead until that is implemented.

**Review by crtrott — CHANGES_REQUESTED** [review]
I don't think we should try and get lvl 1 from the sycl shmem or?

**Comment by masterleinad** [comment]
Superseded by #3763.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:452]: See, #3763 for the pull request corresponding to the new TeamPolicy implementation. Sorry for the confusion!

**Comment by masterleinad** [comment]
> I don't think we should try and get lvl 1 from the sycl shmem or?

Neither I nor @dalg24 was super sure about that. I'll change L1 to map to global memory in https://github.com/kokkos/kokkos/pull/3763 then.

**Pull request closed without merging**